src/js/helpers/util.js

```diff
@@ -60,14 +60,62 @@ var util = {
     var field;
     /*eslint-disable no-cond-assign */
     while (field = nodeIterator.nextNode()) {
-      var val = field.value;
+      var val = field.value.replace(rCRLF, "\r\n");
+      if (field.type === "number") {
+        val = parseFloat(val);
+      }
+      if (field.type === "checkbox") {
+        val = !!val;
+      }
       if (val) {
-        serialized.push({name: field.name, value: val.replace(rCRLF, "\r\n")});
+        serialized.push({name: field.name, value: val});
       }
     }
     /*eslint-enable no-cond-assign */
     return serialized;
   },
+  serializedArrayToDictionary: function (serializedArray = []) {
+    let sanitizePath = function (path) {
+      return path
+        .replace(/^\.|\.$/, "") // leading/trailing dots
+        .replace(/\.\.+/, "."); // multiple dots
+    };
+    let parsePath = function (obj, position, tokens, value) {
+      if (position === tokens.length) {
+        return value;
+      }
+      let token = tokens[position];
+      let matches = token.match(/(\w+)\[(\d*)]$/); // parse array notation
+
+      if (!matches) {
+        if (obj[token] === undefined) {
+          obj[token] = {};
+        }
+        obj[token] = parsePath(obj[token], position + 1, tokens, value);
+      } else {
+        let [key, index] = matches.slice(1);
+        if (obj[key] === undefined) {
+          obj[key] = [];
+        }
+        if (index.length === 0) {
+          index = obj[key].length;
+        }
+        if (obj[key][index] === undefined) {
+          obj[key][index] = {};
+        }
+        obj[key][index] = parsePath(obj[key][index], position + 1, tokens, value);
+      }
+      return obj;
+    };
+
+    let json = {};
+    for (let i = 0; i < serializedArray.length; i++) {
+      const value = serializedArray[i].value;
+      const tokens = sanitizePath(serializedArray[i].name).split(".");
```

Comment:
Yeah, I got it wrong here. `let` only makes sense for `value` (as it holds a reference to primitive type, which means it will be actually treated as "read only") while for `tokens` I should have used a `let`, as it holds a reference to an array, whose contents might actually change without raising an exception. 

The idea was to explicitly signal block scoping here, which in this case makes sense, imho.